tests/regression/patterns/llvm/rhel-81-82/blk_mq_end_request.ll

```diff
@@ -0,0 +1,47 @@
+; Found differences in functions called by blk_mq_end_request
+;
+; arch_atomic_dec differs:
+;   Callstack (snapshots/linux-4.18.0-147.el8):
+;   __blk_mq_end_request at block/blk-mq.c:566
+;   blk_mq_free_request at block/blk-mq.c:557
+;   atomic_dec at block/blk-mq.c:521
+;   arch_atomic_dec at ./include/asm-generic/atomic-instrumented.h:115
+
+;   Callstack (snapshots/linux-4.18.0-193.el8):
+;   __blk_mq_end_request at block/blk-mq.c:562
+;   blk_mq_free_request at block/blk-mq.c:553
+;   atomic_dec at block/blk-mq.c:517
+;   arch_atomic_dec at ./include/asm-generic/atomic-instrumented.h:115
+
+;   Diff:
+;   *************** static __always_inline void arch_atomic_dec(atomic_t *v)
+;   *** 108,110 ***
+;     	asm volatile(LOCK_PREFIX "decl %0"
+;   ! 		     : "+m" (v->counter));
+;     }
+;   --- 108,110 ---
+;     	asm volatile(LOCK_PREFIX "decl %0"
+;   ! 		     : "+m" (v->counter) :: "memory");
+;     }
+
+; Metadata
+!0 = !{ !"pattern-start" }
+!1 = !{ !"pattern-end" }
+
+; Structures
+%struct.custom = type { i32 }
+
+; Functions
+declare void @simpll__inlineasm(...)
+
+define void @diffkemp.old.arch_atomic_dec(%struct.custom*) {
+  %2 = getelementptr inbounds %struct.custom, %struct.custom* %0, i32 0, i32 0
+  call void (...) @simpll__inlineasm(i32* %2, i32* %2), !diffkemp.pattern !0
+  ret void, !diffkemp.pattern !1
+}
+
+define void @diffkemp.new.arch_atomic_dec(%struct.custom*) {
+  %2 = getelementptr inbounds %struct.custom, %struct.custom* %0, i32 0, i32 0
+  call void (...) @simpll__inlineasm(i32* %2, i32* %2), !diffkemp.pattern !0
+  ret void, !diffkemp.pattern !1
+}
```

Comment:
This is not entirely correct: the algorithm should still check that the specific types and the GEP instruction. Either way, I have faced the same dilemma when creating this test but decided to keep the test since (1) pretty sure we can easily extend this with full asm string support (probably using metadata), and (2) in the meantime, it, at the very least, still serves as an example of managing inline assembly (if desired to do so this way).

Reply:
Ok, perhaps just put this into a comment.